File: ccc08s3.py
#ccc08s3
import queue

num_cases = int(input())

for case in range(num_cases):
	width = int(input())
	height = int(input())

	grid = []

	for h in range(width):
		grid.append(list(input()))

	solved = set()
	q = queue.Queue()

	q.put((0, 0, 1))

	broken = False

	while not q.empty():
		x, y, c = q.get()
		solved.add((x, y))

		char  = grid[x][y]

		if x == width - 1 and y == height - 1:
			print(c)
			broken = True
			break

		if char == '-' or char == '+':
			if y - 1 > -1 and grid[x][y - 1] != '*':
				if (x, y - 1) not in solved:
					q.put((x, y - 1, c + 1))

			if y + 1 < height and grid[x][y + 1] != '*':
				if (x, y + 1) not in solved:
					q.put((x, y + 1, c + 1))

		if char == '|' or char == '+':
			if x - 1 > -1 and grid[x - 1][y] != '*':
				if (x - 1, y) not in solved:
					q.put((x - 1, y, c + 1))

			if x + 1 < width and grid[x + 1][y] != '*':
				if (x + 1, y) not in solved:
					q.put((x + 1, y, c + 1))

	if not broken:
			print(-1)

File: ccc08s3.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int main() {
  int numCases, width, height;
  bool broken;

  cin >> numCases;

  for (int c = 0; c < numCases; c++) {
    cin >> width >> height;

    string grid[width][height] = {};

    vector<int[2]> solved;
    queue<int[3]> q = {};
    q.push({0, 0, 1});

    broken = false;

    for (int x = 0; x < width; x++) {
      for (int y = 0; y < height; y++) {
          cin >> grid[x][y];
      }
    }

    while(!q.empty) {
      int val[3] = q.pop();
      int x, y, c = val[0], val[1], val[2];

      solved = insert(solved, {x, y});

      string char = grid[x][y];

    }

    if (!broken) {
        cout << -1 << endl;
    }
  }
}
